ports/views.py

```diff
@@ -78,32 +78,79 @@ def variantlist(request, variant):
-    dependencies = Dependency.objects.filter(port_name_id=port_id)
-    variants = Variant.objects.filter(port_id=port_id)
-
-    builders = Builder.objects.values_list('name', flat=True)
-    build_history = {}
-    for builder in builders:
-        build_history[builder] = BuildHistory.objects.filter(builder_name__name=builder, port_name=name).order_by('-time_start')
+    portname = name
     return render(request, 'ports/portdetail.html', {
-        'port': port,
-        'build_history': build_history,
-        'maintainers': maintainers,
-        'dependencies': dependencies,
-        'variants': variants,
-        'builders_list': builders,
+        'portname': portname,
     })
 
 
+def portdetail_summary(request):
+    try:
+        portname = request.GET['portname']
+        port = Port.objects.get(name=portname)
+        port_id = port.id
+        maintainers = Maintainer.objects.filter(ports__name=portname)
+        dependencies = Dependency.objects.filter(port_name_id=port_id)
+        variants = Variant.objects.filter(port_id=port_id)
+
+        builders = Builder.objects.order_by('-name').values_list('name', flat=True)
+        latest_builds = {}
+        for builder in builders:
+            latest_builds[builder] = BuildHistory.objects.filter(builder_name__name=builder, port_name=portname).order_by(
+                '-time_start').first()
+        return render(request, 'ports/port-detail/summary.html', {
+            'port': port,
+            'latest_builds': latest_builds,
+            'maintainers': maintainers,
+            'dependencies': dependencies,
+            'variants': variants,
+            'builders_list': builders,
+        })
+    except Port.DoesNotExist:
+        return HttpResponse("<br><br><h3>Sorry, the port you requested was not found.</h3><br>")
+
+
+def portdetail_build_information(request):
+    try:
+        portname = request.GET['portname']
+        port = Port.objects.get(name=portname)
+        status = request.GET.get('status', '')
+        builder = request.GET.get('builder_name__name', '')
+        page = request.GET.get('page', 1)
+        builders = Builder.objects.order_by('-name').values_list('name', flat=True)
+        builds = BuildHistoryFilterForPort(request.GET, queryset=BuildHistory.objects.filter(port_name__iexact=portname)).qs
+        paginated_builds = Paginator(builds, 100)
+        try:
+            result = paginated_builds.get_page(page)
+        except PageNotAnInteger:
+            result = paginated_builds.get_page(1)
+        except EmptyPage:
+            result = paginated_builds.get_page(paginated_builds.num_pages)
+
+        return render(request, 'ports/port-detail/build_information.html', {
+            'builds': result,
+            'builder': builder,
+            'builders': builders,
+            'status': status,
+        })
+    except Port.DoesNotExist:
+        return HttpResponse("<br><br><h3>Sorry, the port you requested was not found.</h3><br>")
```

Comment:
No super strong feelings, but as soon as you need that much formatting, it might be time for a template?

Anyway, I'm much happier now than I was with the previous response.

Reply:
Yes, I would also try to cover all such exceptions across the app.